### ISSUE TITLE
[FSDP2] support sync_module_states in FSDP2

### ISSUE THREAD

**Issue by weifengpy** [pytorch/torchtitan#1777]
this is from mistral: both DDP and FSDP1 broadcast weights from rank0 to other ranks, because each rank may have different rng before applying DDP/FSDP1

the best practice for FSDP2 is meta init when applying fully_shard. then call reset_paratmers() on DTensor model.parameters(). DTensor will handle RNG for random ops. 

still. for people from DDP/FSDP1, it nature they ask for sync_module_states equiverlance in FSDP2

**Comment by fegin** [comment]
Why? If all ranks have the same initial random seed? Isn't DTensor going to handle the case? That is how current TorchTitan works. @XilunWu please correct me if I'm wrong.

**Comment by fegin** [comment]
I prefer seed checkpointing instead of letting FSDP2 support this feature is users want a truly identical weight initialization as one GPU. It is general to other parallelisms.

**Comment by rakkit** [comment]
I have a related question there about `DTensor` + `orthogonal` init (we wrap it [here](https://github.com/rakkit/torchtitan/blob/trustllm-0.2/torchtitan/models/inits.py)).
given DTensor dont have `orthogonal` yet, what we do is, firstly, on all ranks, do `torch.nn.init.orthogonal_(temp_tensor, gain=gain)`on normal tensor.

And then we have two possible ways to chunk it to DTensor's local tensor, either 
```
       # Implementation-1: Use `distribute_tensor`
        tensor_data = distribute_tensor(
            temp_tensor, placements=tensor.placements, device_mesh=tensor.device_mesh,
        )
        tensor.copy_(tensor_data)
```
or 
```
        # Implementation-2: Use `DTensor.from_local`
        chunk = tensor.__create_chunk_list__()[0]  
        offs, sizes = chunk.offsets, chunk.sizes  # torch.Size objects
        for dim, (o, s) in enumerate(zip(offs, sizes)):
            temp_tensor = temp_tensor.narrow(dim, o, s)

        tensor.data.to_local().copy_(temp_tensor)
```

We currently prefer the second one because we dont need extra communication for each matrix's initialization. And our question is if all rank's `torch.nn.init.orthogonal_` is "deterministic" to give same results?

**Comment by tianyu-l** [comment]
@rakkit 
I believe you can use the 1st implementation, by setting `src_data_rank=None`.
https://github.com/pytorch/pytorch/blob/349e9e922d6049a76cf5036cdfe846f0f4200e44/torch/distributed/tensor/_api.py#L660C5-L660C18

> And our question is if all rank's torch.nn.init.orthogonal_ is "deterministic" to give same results?

If you start with same random seed across ranks, they should be the same.

Recently there's some improvement on DTensor RNG behavior
PR: https://github.com/pytorch/pytorch/pull/160482
GH issue: https://github.com/pytorch/pytorch/issues/159991

I was thinking maybe you could also do `Replicate` DTensor and then redistribute to `Shard`.

**Comment by weifengpy** [comment]
> Why? If all ranks have the same initial random seed? 

in this case, they are not setting random seed the same across rank, but want fsdp2 to broadcast weights

**Comment by fegin** [comment]
> in this case, they are not setting random seed the same across rank, but want fsdp2 to broadcast weights

Then this seems to be easier and more general to just use seed checkpoint for this purpose? It can apply to even when PP is used, which FSDP2 broadcasting is not enough.

**Comment by weifengpy** [comment]
yes, maybe seed checkpoint is a better option
